### ISSUE TITLE
How to save and load the trained network?

### ISSUE THREAD

**Issue by LeZhengThu** [jaredleekatzman/DeepSurv#7]
I have large amount of data and the training takes quite a lot of time. Therefore I wonder how to save and load the trained network?

**Comment by dareneiri** [comment]
Based on the code in https://github.com/jaredleekatzman/DeepSurv/blob/master/deepsurv/deep_surv.py, you should be able to save and load the model with save_model(), save_weights(), load_weights(), and load_model_from_json().

**Comment by dareneiri** [comment]
@LeZhengThu Here's some additional details until the official documentation is released. I hope this helps. I had to play around with the latest code to figure this out, and this may not be 100% correct but I'm able to save and load a trained network by doing the following: 

1. Uninstall any existing version of deepsurv and reinstall by cloning the repo directly (do not install from pip). You may also need to install TensorFlow as the plotting of CI and loss/validation over epochs is done using TensorBoard. 

1. Note where deepsurv is installed (I use a conda virtual env, with anaconda installed in /usr/local/bin, but my path is `/usr/local/bin/anaconda2/envs/deepsurv/lib/python2.7/site-packages/deepsurv/__init__.py`, and edit `__init__.py` so it is the following:
    ```python
    from .deep_surv import DeepSurv
    from .deepsurv_logger import DeepSurvLogger
    from .deepsurv_logger import TensorboardLogger
    from .deep_surv import load_model_from_json
    from .viz import plot_log
    import datasets
    ```

1. Then I import the following:
    ```python
    from deepsurv import DeepSurv
    from deepsurv import DeepSurvLogger
    from deepsurv import TensorboardLogger
    ```

1. Once you define your hyperparameters, you'll need to modify the training function:
    ```python
    network.restored_update_params = None 
    logger = TensorboardLogger(name = 'DeepSurvLogger', logdir = '/home/spark/test_run2')
    log =  network.train(dataset_TRAIN, dataset_VALID, n_epochs = 500, logger = logger) 
    ```
1.  You can save the model like so: 
    ```python
    network.save_model("deep_surv_model_500_epochs", weights_file ="deep_surv_weights_500_epochs")
    ```
This saves the hyperparameters and weights in my working directory. 


To load the saved hyperparameters and weights: 
```python
previous_model = load_model_from_json(model_fp = 'deep_surv_model_500_epochs' , weights_fp =     'deep_surv_weights_500_epochs')
restored_update_params = previous_model
```

Then you can use: 
```python
previous_model.get_concordance_index(**dataset_TEST)
```

And predict the risk of the dataset without running the model: 
```python
predict_risk = previous_model.predict_risk(risk_array)
print(predict_risk[0:10,])
```

**Comment by dareneiri** [comment]
@jaredleekatzman I can save the model and its weights using save_model(), and then load it in a different file with load_model_from_json(). Using the same test data I used when building the model, I get different CI value (difference of 0.10) from the model that was saved and loaded. 

Is this something I should be expecting until your upcoming release (a known issue)?

**Comment by dareneiri** [comment]
@jaredleekatzman Can you quickly tell me if the loading/saving functionality is working properly, or if it's something you are still working on? I can load and save, but when I test the model with loaded parameters and weights, I don't think the weights are applying -- it's as if the model predicts without any weights, just the parameters. 

For example, once I load the data I do the following: 
```python
previous_model = load_model_from_json(model_fp = 'deep_surv_model_300ep' , weights_fp = 'deep_surv_weights_300ep')
print(previous_model.__dict__)

{'partial_hazard': Elemwise{exp,no_inplace}.0, 'scale': <GpuArrayType<None>(float32, vector)>, 'standardize': True, 'network': <lasagne.layers.dense.DenseLayer object at 0x7f750a8cafd0>, 'n_in': 15, 'learning_rate': 0.0001, 'hyperparams': {'standardize': True, 'hidden_layers_sizes': None, 'activation': u'rectify', 'L1_reg': 0, 'dropout': 0, 'n_in': 15, 'lr_decay': 0, 'L2_reg': 0.01, 'learning_rate': 0.0001, 'momentum': 0.9, 'batch_norm': False}, 'restored_update_params': [array([[-0.80010158],
       [-0.12430875],
       [ 0.07427295],
       [-0.40899086],
       [-1.48216331],
       [-0.73171037],
       [-0.94763297],
       [-0.23204659],
       [-0.60043931],
       [-0.94336671],
       [-0.76649785],
       [ 0.33748096],
       [ 0.51455891],
       [-0.9143433 ],
       [ 0.41974837]], dtype=float32), array([-0.56789464], dtype=float32), array([[ 0.],
       [ 0.],
       [ 0.],
       [ 0.],
       [ 0.],
       [ 0.],
       [ 0.],
       [ 0.],
       [ 0.],
       [ 0.],
       [ 0.],
       [ 0.],
       [ 0.],
       [ 0.],
       [ 0.]], dtype=float32), array([ 0.], dtype=float32)], 'lr_decay': 0, 'params': [W, b], 'offset': <GpuArrayType<None>(float32, vector)>, 'L1_reg': 0, 'X': x, 'E': e, 'hidden_layers': [<lasagne.layers.special.BiasLayer object at 0x7f750a8ca0d0>, <lasagne.layers.special.ScaleLayer object at 0x7f750a8caf50>, <lasagne.layers.dense.DenseLayer object at 0x7f750a8cafd0>], 'momentum': 0.9, 'L2_reg': 0.01}
```
And this looks correct. 

But when I predict the risk on the same exact dataset, the values I get are very different (and the risk_array is in the same order). The CI is also different, but I would expect them to be the same for the same dataset. 

```python
predict_risk = previous_model.predict_risk(risk_array)
```


thank you!!

**Comment by jaredleekatzman** [comment]
I haven't tested the saving & loading functionalities thoroughly, so you are most likely encountering a bug

**Comment by dareneiri** [comment]
@jaredleekatzman Thanks for the quick reply!

**Comment by dareneiri** [comment]
@jaredleekatzman I reviewed the function `load_weights()` and I found that in my case, when printing `fp` in  `load_all_keys()`, only the updates were loading. So it seemed that there were no weights being saved when using `save_model()`

As a quick fix, I forked your master and within `save_list_by_idx()` I added:
```python
numpy.savez('model.npz', *lasagne.layers.get_all_param_values(self.network))
```

And to load model.npz, I added this within` load_all_keys()`:
```python
        with numpy.load('model.npz') as f:
            param_values = [f['arr_%d' % i] for i in range(len(f.files))]
        lasagne.layers.set_all_param_values(self.network, param_values)
```

Now when I use deepsurv's method of saving and loading the trained model, I get the same CI value and risk values for each subject. 

I hope this helps you -- I am sure you are busy working on your next release!
